functorch/functorch/experimental/cond.py

```diff
@@ -0,0 +1,133 @@
+    curr_mode = torch._C._get_torch_dispatch_mode()
+    torch._C._set_torch_dispatch_mode(mode)
+    try:
+        yield
+    finally:
+        torch._C._set_torch_dispatch_mode(curr_mode)
+
+
+def trace_cond(proxy_mode, func_overload, pred, true_fn, false_fn, operands):
+    def _unwrap_proxy(e):
+        return get_proxy_slot(e, proxy_mode.tracer, e, lambda e: e.proxy)
+
+    assert isinstance(operands, list), "Cond operands must be a list of tensors"
+    assert all(isinstance(o, torch.Tensor) for o in operands), "Cond operands must be a list of tensors"
+
+    true_graph = get_isolated_graphmodule(true_fn, operands, {})
+    false_graph = get_isolated_graphmodule(false_fn, operands, {})
+
+    # There are probably better ways - I know that create_arg has some self incrementing name
+    # magic to it, but since we explicitly have to get the name for register_module,
+    # I was not sure how to do that. This kinda simulates it.
+    next_name = None
+    i = 0
+    while not next_name:
+        candidate = f"true_graph_{i}"
+        if hasattr(proxy_mode.tracer.root, candidate):
+            i += 1
+        else:
+            next_name = candidate
+
+    true_name = next_name
+    false_name = f"false_graph_{i}"
+    assert(not hasattr(proxy_mode.tracer.root, false_name))
+
+    proxy_mode.tracer.root.register_module(true_name, true_graph)
+    proxy_mode.tracer.root.register_module(false_name, false_graph)
+
+    # This is not amazing.
+    # However, if we have nested operators that have a call_function
+    # in their graph that is not a torch op (ex: see conditional below, nested cond)
+    # we cannot get metadata for it from just looking at out vars.
+    # The reason is that an operation on the output of such an op is not
+    # evalauted as a torch.Tensor.
+    # So we execute the real true and false fn here and compare metadata
+    true_result = true_graph(operands)
+    false_result = false_graph(operands)
+
+    def recursive_compare_same(a, b):
+        assert(type(a) == type(b))
+        if isinstance(a, torch.Tensor):
+            assert(a.dtype == b.dtype)
+            assert(a.size() == b.size())
+            assert(a.stride() == b.stride())
+            assert(a.device == b.device)
+        elif isinstance(a, (list, tuple)):
+            assert(len(a) == len(b))
+            for i in range(0, len(a)):
+                recursive_compare_same(a[i], b[i])
+
+    recursive_compare_same(true_result, false_result)
+
+    args = (pred, true_graph, false_graph, [operands])
+
+    proxy_args = pytree.tree_map(_unwrap_proxy, args)
+
+    return proxy_mode.tracer.create_proxy('call_function', func_overload, proxy_args, {},
+        name="conditional")
+
+
+def cond_dense(pred, true_fn, false_fn, operands):
+    mode = torch._C._get_torch_dispatch_mode()
+    assert (mode is None), "Mode should never be enabled for CPU key"
+    if pred:
+        return true_fn(*operands)
+    else:
+        return false_fn(*operands)
```

Comment:
What happened to the `*operands`? The API for cond was previously `cond(pred, true_fn, false_fn, *operands)`, right?

Reply:
Yes, @ezyang said it might be easier and cleaner for the cond API to be `cond, pred, true_fn, false_fn, [tensor]` and I agree, that makes it cleaner.

Reply:
discused offline